src/backend/distributed/utils/shard_pruning.c

```diff
@@ -0,0 +1,1239 @@
+ */
+#include "postgres.h"
+
+#include "distributed/shard_pruning.h"
+
+#include "access/nbtree.h"
+#include "catalog/pg_am.h"
+#include "catalog/pg_collation.h"
+#include "catalog/pg_type.h"
+#include "distributed/metadata_cache.h"
+#include "distributed/multi_planner.h"
+#include "distributed/multi_join_order.h"
+#include "distributed/multi_physical_planner.h"
+#include "distributed/shardinterval_utils.h"
+#include "distributed/pg_dist_partition.h"
+#include "distributed/worker_protocol.h"
+#include "nodes/nodeFuncs.h"
+#include "optimizer/clauses.h"
+#include "utils/catcache.h"
+#include "utils/lsyscache.h"
+#include "utils/memutils.h"
+
+/*
+ * A pruning instance is a set of ANDed constraints on a partition key.
+ */
+typedef struct PruningInstance
+{
+	/*
+	 * This constraint never evaluates to false, i.e. pruning does not have to
+	 * be performed.
+	 */
+	bool alwaysFalse;
+
+	/*
+	 * Constraints on the partition column value.  If multiple values are
+	 * found the more restrictive one should be stored here.  Even in case of
+	 * a hash-partitioned table, actual column-values are stored here, *not*
+	 * hashed values.
+	 */
+	Const *lessConsts;
+	Const *lessEqualConsts;
+	Const *equalConsts;
+	Const *greaterEqualConsts;
+	Const *greaterConsts;
+
+	/*
+	 * Constraint using a pre-hashed column value. The constant will store the
+	 * hashed value, not the original value of the restriction.
+	 */
+	Const *hashedEqualConsts;
+
+	/*
+	 * Types of constraints not understood.  We could theoretically try more
+	 * expensive methods of pruning if any such restrictions are found.
+	 *
+	 * TODO: any actual use for this? Right now there seems little point.
+	 */
+	List *otherRestrictions;
+
+	/*
+	 * Has this PruningInstance been added to
+	 * ClauseWalkerContext->pruningInstances? This is not done immediately,
+	 * but the first time a constraint (independent of us being able to handle
+	 * that constraint) is found.
+	 */
+	bool added;
+
+	/*
+	 * When OR clauses are found, the non-ORed part (think of a < 3 AND (a > 5
+	 * OR a > 7)) of the expression is stored in one PruningInstance which is
+	 * then copied for the ORed expressions.  The original is marked as
+	 * isPartial, to avoid it being used for pruning.
+	 */
+	bool isPartial;
+
+	/*
+	 * Does this instance contain any prunable expressions.
+	 */
+	bool hasValidConstraint;
+} PruningInstance;
+
+
+/*
+ * Partial instances that need to be finished building.  This is used to
+ * collect all ANDed restrictions, before looking into ORed expressions.
+ */
+typedef struct PendingPruningInstance
+{
+	PruningInstance *instance;
+	Node *continueAt;
+} PendingPruningInstance;
+
+
+/*
+ * Data necessary to perform a single PruneShards().
+ */
+typedef struct ClauseWalkerContext
+{
+	Var *partitionColumn;
+	char partitionMethod;
+
+	/* ORed list of pruning targets */
+	List *pruningInstances;
+
+	/*
+	 * Partially built PruningInstances, that need to be completed. XXX
```

Comment:
Think I wanted to expand the comment a bit, or rather reference a comment explaining why partial instances are needed.